Documentation/DesignDocs/corewcfqueue-design.md

```diff
@@ -0,0 +1,190 @@
+```csharp
+public class QueueMiddleWareBuilderExtension
+{
+    internal const string InvokeAsyncMethodName = "InvokeAsync";
+    public static IQueueMiddlewareBuilder UseMiddleware<TMiddleware>(this IQueueMiddlewareBuilder app, params object[] args)
+    {
+         return app.UseMiddleware(typeof(TMiddleware), args);
+    }
+    
+    public static IQueueMiddlewareBuilder UseMiddleware(this IQueueMiddlewareBuilder app, Type middleware, params object[] args)
+    {
+        //TODO
+     }
+    }
+```
+#### QueueServiceWebHostBuilderExtension
+- QueueServiceWebHostBuilderExtension is the WebHostBuilder extension which stiches all the middleware pipeline and ```QueuePollingService``` which will be called during service startup for queuing mechanism to work. e:g:- 
+```csharp
+internal sealed class Startup
+{
+    public void ConfigureServices(IServiceCollection services)
+    {
+      services.AddServiceModelServices();
+      services.UseCoreWCFQueue();
+ }
+ ```
+ - It first keeps an instance of IQueueMiddlewareBuilder for chaining all the middleware components to be used
+ - Afterwards it initalizes the ```QueueHandShakeMiddleWare``` and ```QueuePollingService```.
+#### QueueHandShakeMiddleWare
+- The QueueHandShakeMiddleWare has 2 methods, ```InitQueueMiddleWare``` and ```Build```.
+- The InitQueueMiddleWare initalizes middlewares to use and is called from constructor
+- Build method builds the middleware pipeline and returns delegate.
+#### QueueBaseBindingElement , QueueTransportContext and IQueueTransport
+- IQueueTransport holds the signature for method ```ReceiveQueueMessageContextAsync```. 
+```csharp
+   public interface IQueueTransport
+    {
+        Task<QueueMessageContext> ReceiveQueueMessageContextAsync(CancellationToken cancellationToken);
+    }
+```
+- QueueBaseBindingElement is an abstract class with ```CreateQueueTransport``` as abstract method. Any concrete QueueTransport part of the CoreWCF, needs to implement and provide a concrete implementation of IQueueTransport.
+```csharp
+    public abstract class QueueBaseBindingElement : BindingElement
+    {
+        protected QueueBaseBindingElement()
+        {
+        }
+
+        protected QueueBaseBindingElement(QueueBaseBindingElement elementToBeCloned)
+            : base(elementToBeCloned)
+        {
+        }
+
+        public abstract IQueueTransport CreateQueueTransport(BindingContext context);
+        public virtual int MaxPendingReceives { get { return 1; } }
+    }
+ ```
+- The concrete implementation of IQueueInternalTransport will be used by CoreWCF to retrieve messages from the queue storage service. 
+- When ReceiveQueueMessageContextAsync method is called, it should return a a concrete implementation of QueueMessageContext object.
```

Comment:
Weren't we going to try to have QueueMessageContext be usable for many scenarios as is and the only reason to derive it would be if a particular implementation needs to do something different?

Reply:
do you want me to add this ?